powerfulseal/policy/action_clone.py

```diff
@@ -0,0 +1,331 @@
+
+# Copyright 2017 Bloomberg Finance L.P.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import requests
+import time
+import kubernetes.client
+
+from powerfulseal import makeLogger
+
+from ..metriccollectors.stdout_collector import StdoutCollector
+from .action_abstract import ActionAbstract
+
+DEFAULT_TOXIPROXY_IMAGE = "docker.io/shopify/toxiproxy:2.1.4"
+DEFAULT_IPTABLES_IMAGE = "gaiadocker/iproute2:latest"
+
+
+class DeleteDeploymentAction():
+  def __init__(self, name, namespace, k8s_inventory, logger=None):
+    self.name = name
+    self.namespace = namespace
+    self.k8s_inventory = k8s_inventory
+    self.logger = logger or makeLogger(__name__)
+
+  def execute(self):
+    try:
+      response = self.k8s_inventory.k8s_client.delete_deployment(
+        namespace=self.namespace,
+        name=self.name,
+      )
+      self.logger.debug("Response %s", response)
+      self.logger.info("Clone deployment deleted successfully: %s in %s", self.name, self.namespace)
+    except:
+      self.logger.exception("Error deleting clone deployment: %s in %s", self.name, self.namespace)
+      return False
+
+class ActionClone(ActionAbstract):
+
+  def __init__(self, name, schema, k8s_inventory, logger=None, metric_collector=None):
+    self.name = name
+    self.schema = schema
+    self.k8s_inventory = k8s_inventory
+    self.logger = logger or makeLogger(__name__, name)
+    self.metric_collector = metric_collector or StdoutCollector()
+    self.cleanup_actions = []
+
+  def get_cleanup_actions(self):
+    return self.cleanup_actions
+
+  def get_source_schema(self, source):
+    # currently, only deployments are supported
+    source_deployment = source.get("deployment")
+    deployment = self.k8s_inventory.k8s_client.get_deployment(
+      name=source_deployment.get("name"),
+      namespace=source_deployment.get("namespace"),
+    )
+    return deployment
+
+  def modify_labels(self, body, update_labels):
+    """
+      Modifies labels and deployment selectors as required by the schema
+    """
+    for label_modifier in self.schema.get("labels", []):
```

Comment:
why are you iterating on the labels? can you not update the labels after consolidating all the labels so you dont have to make multiple calls to get service.

Reply:
I'm not sure what that sentence means. I'm not iterating on labels... I'm iterating on the label modifiers. You can specify multiple label modifiers, that can potentially override bits of each other. Example: copy labels over from service X matchers, and then add static labels of my choosing.